### PR TITLE
feat: Automation account credentials - New feature

### LINKED ISSUE #1443: [AVM Module Issue]: Automation account credentials
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Issue Type?

Feature Request

### Module Name

avm/res/automation/automation-account

### (Optional) Module Name if not listed above

_No response_

### (Optional) Module Version

_No response_

### Description

It doesn't seem like adding credential resources to an Automation Account is supported by the AVM. I'm having an extremely difficult time determining how to combin usage of the AA AVM with a manual declaration of a credential resource. Wondering if it's even possible.

### (Optional) Correlation Id

_No response_

### PR DESCRIPTION
**Pull request by elanzel**
## Description

Fixes #1443
## Pipeline Reference
| Pipeline |
| -------- |
|       [avm.res.automation.automation-account](https://github.com/elanzel/bicep-registry-modules/actions/workflows/avm.res.automation.automation-account.yml)   |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [x] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

**Review by segraef — DISMISSED** [review]
lgtm

File: avm/res/automation/automation-account/credential/main.bicep [hunk 1]
```diff
@@ -0,0 +1,42 @@
+metadata name = 'Automation Account Credentials'
+metadata description = 'This module deploys Azure Automation Account Credentials.'
+metadata owner = 'Azure/module-maintainers'
+
+@description('Required. Name of the Automation Account credential.')
+param credentialName string
```
Comment (AlexanderSehr) [inline avm/res/automation/automation-account/credential/main.bicep:6]: It's usually a convention that there's always just a `name` parameter as it should be clear that a `name` parameter for a `credential` module implies that you're setting the credential's name :) 

```suggestion
param name string
```

Comment (AlexanderSehr) [reply avm/res/automation/automation-account/credential/main.bicep:6]: Please note, this must be updated in several other locations too and requires a regeneration of the files

File: avm/res/automation/automation-account/main.bicep [hunk 2]
```diff
@@ -18,6 +18,9 @@ param skuName string = 'Basic'
 @description('Optional. The customer managed key definition.')
 param customerManagedKey customerManagedKeyType
 
+@description('Optional. List of credentials to be created in the automation account.')
+param credentials array = []
```
Comment (AlexanderSehr) [inline avm/res/automation/automation-account/main.bicep:22]: We need a UDT for this to not expose the credential's secret

`arrays` don't support the `@secure()` flag - so we need to break it down into the User-Defined-Type and just like you configured it in the sub-module declare the actual secret as `@secure()`

File: avm/res/automation/automation-account/credential/main.bicep [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+@description('Required. The user name associated to the credential.')
+param userName string
+
+@description('Required. Password of the credential.')
+@secure()
+param password string
+
+@description('Optional. Description of the credential.')
+param credentialDescription string = ''
```
Comment (AlexanderSehr) [inline avm/res/automation/automation-account/credential/main.bicep:19]: Same comment as for the `name`. The `credential` just seems redundant 😉

File: avm/res/automation/automation-account/main.bicep [hunk 4]
```diff
@@ -194,6 +193,19 @@ resource automationAccount 'Microsoft.Automation/automationAccounts@2022-08-08'
   }
 }
 
+module automationAccount_credentials 'credential/main.bicep' = [
+  for (credential, index) in (credentials ?? []): {
+    name: '${uniqueString(deployment().name, location)}-AutomationAccount-Credential-${index}'
+    params: {
+      credentialName: credential.credentialName
+      credentialDescription: credential.credentialDescription
```
Comment (AlexanderSehr) [inline avm/res/automation/automation-account/main.bicep:201]: Unless you want to make it mandatory 😉 
```suggestion
      credentialDescription: credential.?description
```

File: avm/res/automation/automation-account/credential/main.bicep (hunk 1 above)
Comment (elanzel) [reply avm/res/automation/automation-account/credential/main.bicep:6]: parameter name updated

File: avm/res/automation/automation-account/credential/main.bicep (hunk 3 above)
Comment (elanzel) [reply avm/res/automation/automation-account/credential/main.bicep:19]: description parameter updated

File: avm/res/automation/automation-account/main.bicep (hunk 4 above)
Comment (elanzel) [reply avm/res/automation/automation-account/main.bicep:201]: updated

File: avm/res/automation/automation-account/credential/main.bicep (hunk 1 above)
Comment (AlexanderSehr) [reply avm/res/automation/automation-account/credential/main.bicep:6]: Hey @elanzel, I suspect you may have forgotten to push your commit 😉

File: avm/res/automation/automation-account/credential/main.bicep (hunk 3 above)
Comment (elanzel) [reply avm/res/automation/automation-account/credential/main.bicep:19]: unluckily not: introducing the user defined type

File: avm/res/automation/automation-account/credential/main.bicep (hunk 1 above)
Comment (elanzel) [reply avm/res/automation/automation-account/credential/main.bicep]: updated

File: avm/res/automation/automation-account/credential/main.bicep (hunk 3 above)
Comment (elanzel) [reply avm/res/automation/automation-account/credential/main.bicep]: implemented

File: avm/res/automation/automation-account/credential/main.bicep [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+type credentialType = {
+  @description('Required. Name of the Automation Account credential.')
+  name: string
+
+  @description('Required. The user name associated to the credential.')
+  userName: string
+
+  @description('Optional. Password of the credential.')
+  @secure()
+  password: string
```
Comment (AlexanderSehr) [inline avm/res/automation/automation-account/credential/main.bicep:50]: Is it optinoal as the description suggests? If so, you'd need to make it nullable.

File: avm/res/automation/automation-account/credential/main.bicep [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      description: credential.?description ?? ''
+    }
+  }
+]
+
+@description('The id of the credential associated to the automation account.')
+output resourceId string = automationAccount_credential[0].id
+
+@description('The name of the credential associated to the automation account.')
+output name string = automationAccount_credential[0].name
```
Comment (AlexanderSehr) [inline avm/res/automation/automation-account/credential/main.bicep:31]:
```suggestion
@description('The resource IDs of the credentials associated to the automation account.')
#disable-next-line outputs-should-not-contain-secrets // Does not return the secret, but just the ID
output resourceId array = [for index in range(0, length(credentials)): automationAccount_credential[index].id]

@description('The names of the credentials associated to the automation account.')
#disable-next-line outputs-should-not-contain-secrets // Does not return the secret, but just the name
output name array = [for index in range(0, length(credentials)): automationAccount_credential[index].name]
```

File: avm/res/automation/automation-account/main.bicep [hunk 7]
```diff
@@ -18,6 +18,10 @@ param skuName string = 'Basic'
 @description('Optional. The customer managed key definition.')
 param customerManagedKey customerManagedKeyType
 
+import { credentialType } from 'credential/main.bicep'
```
Comment (AlexanderSehr) [inline avm/res/automation/automation-account/main.bicep:21]: I'm curious, it seems this is now GA? Looks neat
cc: @eriqua for awareness

File: avm/res/automation/automation-account/credential/main.bicep (hunk 5 above)
Comment (elanzel) [reply avm/res/automation/automation-account/credential/main.bicep:50]: Ops, no it's a mistake :-)

Comment (elanzel) [reply avm/res/automation/automation-account/credential/main.bicep:52]: Updated

**Comment by elanzel** [comment]
@AlexanderSehr if you agree, now it's ready for a merge

File: avm/res/automation/automation-account/main.bicep (hunk 7 above)
Comment (AlexanderSehr) [reply avm/res/automation/automation-account/main.bicep:21]: Seems to be GA since `0.25.3`: https://github.com/Azure/bicep/pull/13150

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**